Docs: workarounds for router not resetting state (fixes #35138) (#35651)
vercel/next.js: docs/api-reference/next/router.md

}
```

#### Resetting state after navigation

> When navigating to the same page in Next.js, the page's state **will not** be reset by default:

Reviewer: ```suggestion
When navigating to the same page in Next.js, the page's state **will not** be reset by default as react does not unmount unless the parent component has changed. 
```

```jsx

> // pages/[pageTitle].jsx

Reviewer: ```suggestion
// pages/[slug].js
```

import Link from "next/link"

> import { useState } from "react"

Reviewer: ```suggestion
import { useState } from "react"
import { useRouter } from "next/router"
```

> export const getServerSideProps = async ({ params }) => ({
>   props: {
>     pageTitle: params["pageTitle"]
>   }
> })
> 
> export default function Page({ pageTitle }) {
>   const [title] = useState(pageTitle)
>   return (
>     <div>
>       <h1>{title}</h1>
>       <Link href="/one">
>         <a>one</a>
>       </Link>{" "}
>       <Link href="/two">
>         <a>two</a>
>       </Link>
>     </div>
>   )
> }

Reviewer: ```suggestion
export default function Page(props) {
  const router = useRouter()
  const [count, setCount] = useState(0)
  return (
    <div>
      <h1>Page: {router.query.slug}</h1>
      <p>Count: {count}</p>
      <button onClick={() => setCount(count + 1)}>Increase count</button>
      <Link href="/one">
        <a>one</a>
      </Link>{" "}
      <Link href="/two">
        <a>two</a>
      </Link>
    </div>
  )
}
```

```

> Above, navigating between `/foo` and `/bar` **will not** change the page title. The `useState` is initialized for the first `Page` component, and not re-initialized when navigating to a different `Page` instance. This is because the top-level React component, `Page`, is the same.

Reviewer: ```suggestion
In the above example, navigating between `/one` and `/two` **will not** reset the count . The `useState` is maintained between renders because the top-level React component, `Page`, is the same.
```

If you do not want this behavior, you have a couple of options:

1. Manually ensure each state is updated using `useEffect`. Above, that could look like:

   ```jsx

>    useEffect(() => {
>      setTitle(pageTitle)
>    }, [pageTitle])

Reviewer: ```suggestion
   useEffect(() => {
     setCount(0)
   }, [router.query.slug])
```

>    ```

Reviewer 2: I still don't think this should be our recommendation. If you receive `pageTitle` as a prop, just simply use it in the component, without syncing it into a state.

> 2. Use a React `key` to [tell React to remount the component](https://kentcdodds.com/blog/understanding-reacts-key-prop#:~:text=This%20allows%20you%20to%20return%20the%20exact%20same%20element%20type%2C%20but%20force%20React%20to%20unmount%20the%20previous%20instance%2C%20and%20mount%20a%20new%20one.). To do this for all pages, you can use a custom app:

Reviewer: ```suggestion
2. Use a React `key` to [tell React to remount the component](https://reactjs.org/docs/lists-and-keys.html#keys). To do this for all pages, you can use a custom app:
```

>    ```jsx
>    // pages/_app.jsx
>    import { useRouter } from "next/router"
> 
>    export default function MyApp({ Component, pageProps }) {
>      const router = useRouter()
>      return <Component key={router.asPath} {...pageProps} />
>    }
>    ```

Reviewer: ```suggestion
```jsx
// pages/_app.js
import { useRouter } from "next/router"

export default function MyApp({ Component, pageProps }) {
  const router = useRouter()
  return <Component key={router.asPath} {...pageProps} />
}
```
```
